scripts/linters/pylint_extensions_test.py

```diff
@@ -2915,3 +2915,87 @@ def test_usage_of_is_not_on_none_does_not_add_message(self):
         compare_node = if_node.test
         with self.checker_test_object.assertNoMessages():
             self.checker_test_object.checker.visit_compare(compare_node)
+
+
+class NonTestFilesFunctionNameCheckerTests(unittest.TestCase):
+
+    def setUp(self):
+        super(NonTestFilesFunctionNameCheckerTests, self).setUp()
+        self.checker_test_object = testutils.CheckerTestCase()
+        self.checker_test_object.CHECKER_CLASS = (
+            pylint_extensions.NonTestFilesFunctionNameChecker)
+        self.checker_test_object.setup_method()
+
+    def test_function_def_for_test_file_with_test_only_adds_no_msg(self):
+        def_node = astroid.extract_node(
+            """
+            def test_only_some_random_function(param1, param2):
+                pass
+            """
+        )
+        temp_file = tempfile.NamedTemporaryFile(suffix='_test')
```

Comment:
Instead of creating node can we create tempfile with code? (that way we don't have to explicitly put change the filename on line 2938*)

Reply:
I tried doing that only first i.e write the code into a temp file, and create the node by processing it, but I did not succeed in that, so this is the best I had.

Below I will describe in detail what I tried. I saw the following as some possible solutions to the problem you raised:
- **Solution 1** (will work, though less preferred): I looked at the implementation of [FunctionArgsOrderCheckerTests](https://github.com/oppia/oppia/blob/f6b91cd68a3cfdb75b686e12c55cc19552957eca/scripts/linters/pylint_extensions_test.py#L1907) and [RestrictedImportCheckerTests](https://github.com/oppia/oppia/blob/f6b91cd68a3cfdb75b686e12c55cc19552957eca/scripts/linters/pylint_extensions_test.py#L1952). Here we do not create any actual file or write any code into it, and simply specify the desired test value for `modnode.name`.
    - **Pros**: Tested this approach, will work.
    - **Cons**: I believe "bypassing" the actual mechanism in tests can prove to be fatal in future. In my case it totally depends on what the file name is, and I expect a better solution.
-  **Solution 2**: Other tests have a general pattern like the one followed by `DocstringParameterCheckerTests`'s [test_no_newline_below_class_docstring](https://github.com/oppia/oppia/blob/f6b91cd68a3cfdb75b686e12c55cc19552957eca/scripts/linters/pylint_extensions_test.py#L321) i.e create `astroid.scoped_nodes.Module` object, write code into file and put its filename in Module object, and call the function. I had the written code for my tests on similar lines, but it failed (for the test `test_function_def_for_non_test_file_with_test_only_adds_msg`, you can find **code that I wrote and its run output** [here](https://gist.github.com/sajalasati/c9a77a4af9ff335c7dca6ebe30637bf4)). I am unable to figure out why the actual generated message is empty for all the 4 cases I have written. Though I did some inspection, that might sound interesting.
    - I found `DocstringParameterCheckerTests` also has other tests that perform checks on functions i.e who are using `visit_functiondef` method (like [test_malformed_args_section](https://github.com/oppia/oppia/blob/f6b91cd68a3cfdb75b686e12c55cc19552957eca/scripts/linters/pylint_extensions_test.py#L564)), and for some reason use this method for writing into the files - they create an `extract_node` object, and then do the checking (see [here](https://github.com/oppia/oppia/blob/f6b91cd68a3cfdb75b686e12c55cc19552957eca/scripts/linters/pylint_extensions_test.py#L564-L582)).
    - I checked this is True for all the tests using `visit_functiondef` in pylint_extentions_test file. All of them use this method only (i.e they never write code into a file). I know not all tests require to write code into a temp file, but when they are doing so for one kind of node (say Class), they should be doing it for other kinds of node (say Functions) too. I just gave an example of this above.

To be honest I did not understand how tests are working using the second approach, because nowhere we are actually processing the file content, hence I am unable to debug **my code** ([gist link](https://gist.github.com/sajalasati/c9a77a4af9ff335c7dca6ebe30637bf4)). How do you advise to move ahead?

On a side note, I have ran this PR implementation of the pylint_extension on a sample file for all test cases, and it works correctly. Though as described above, I am unable to replicate the same in tests.

Reply:
As per offline discussion with @DubeySandeep, proceeding with Solution 2 has some issues with the Pylint, which I am unable to resolve at the moment, so I am proceeding with a Solution1 approach in this PR to unblock myself. 

But also since solution 1 is not a good way of testing out our linters (we should be reading the code from the file and not be using a proxy mechanism for that), I have created an issue #11617 to look for a solution, and to change this and other linter checks to use Solution2 styled approach as much as possible.

Reply:
I took another pass on solution 2 and found that the logic for the expected test structure is correct but there was some implementation issue. @sajalasati is working on finding the correct way to write tests for the linters.

Reply:
Hi @DubeySandeep! I am finding myself completely clueless in understanding how `pre_commit_linter` uses the checks implemented by `pylint_extensions`. I can see `pre_commit_linter` imports `python_linter` [here](https://github.com/oppia/oppia/blob/develop/scripts/linters/pre_commit_linter.py#L72). 

Could you please help me by guiding me to location in code where pylint_extension checks are included, so that I can understand how I can emulate the same behaviour in my tests. (Sorry if I am missing some point from our previous discussion on this)

Reply:
@sajalasati, pylint_extensions are some custom checks for the pylint, we import these custom check files through `.pylintrc` and the python_linter runs the pylint and pylint manages running the custom checks files mentioned in `.pylintrc` file. Is this helpful? :)

Reply:
@DubeySandeep Thanks!
One way is - we call `ThirdPartyPythonLintChecksManager.lint_py_files()` in `python_linter.py` in our tests and set its `self.files_to_lint` to the name of the temp file which will contain the test code. 
The problem with this is that `lint_py_files()` runs all the custom checks that are mentioned in `.pylintrc` file, providing no control over the specific Pylint extensions that we would want to run i.e all the extensions would be run on that temp file. We could create a template test file in our codebase with basic code (like `VALID_TEST_FILEPATH`), and then replace things using mocks in actual tests, but running all the checks will take unnecessary time. What do you think about this?

Reply:
As per offline discussion with @DubeySandeep, finally proceeding with Solution 1 approach written above, because we are focusing on testing the `visit_functiondef` method of the newly added linter class as the pylint runner manages populating the checker objects with correct data and calling the visit_<node> method as needed.